submatrix/orbit_auth.py

```diff
@@ -0,0 +1,62 @@
+import sys
+import pycurl
+import re
+from synapse import module_api
+
+def check_credentials(username, password):
+    auth_url = 'http://orbit:9098/mail_auth'
+    headers = [
+        f'Auth-User: {username}',
+        f'Auth-Pass: {password}',
+        'Auth-Protocol: smtp',
+        'Auth-Method: plain',
+    ]
+
+    c = pycurl.Curl()
+    c.setopt(c.URL, auth_url)
+    c.setopt(c.HTTPHEADER, headers)
+
+    try:
+        c.perform()
+        res = c.getinfo(pycurl.RESPONSE_CODE)
+        return res == 200
+    except pycurl.error as e:
+        print(f'error: request failed: {e}', file=sys.stderr)
+        return False
+    finally:
+        c.close()
+
+
+class OrbitAuthProvider:
+    def __init__(self, _, api):
+
+        self.api = api
+
+        api.register_password_auth_provider_callbacks(
+                auth_checkers={("m.login.password", ("password",)): self.handle_login})
+
+    async def handle_login(self, input_username, login_type, login_dict):
+        if login_type != "m.login.password":
+            return None
+
+        # We assume the user specifies their username in the canonical format
+        # of "@<username>:<domain>", however only the username is the Orbit ID
+        # so we must extract it before invoking the Orbit password validation
+        # mechanism provided by radius for mail authentication
+        if not (raw_username_match := re.search(r'@([a-zA-Z0-9._-]+):.*$', input_username)):
```

Comment:
I don't like rolling our own logic here, perhaps UserID.from_string from the API would be better. Also we should check the hostname (easy to do we if use their api).